User ldap dynamic groups support

alexweirig opened the pull request:
Added support for dynamic LDAP groups based on a filter attribute (e.g. memberURL).

Commit:
Support for dynamic LDAP groups

Added support for dynamic LDAP groups based on a filter attribute (e.g. memberURL).

Commit:
Support for dynamic LDAP groups

Added support for dynamic LDAP groups based on a filter attribute (e.g. memberURL).

Commit:
Support for dynamic LDAP groups

Added support for dynamic LDAP groups based on a filter attribute (e.g. memberURL).

Commit:
Update group_ldap.php

Commit:
Support for dynamic LDAP groups

Added support for dynamic LDAP groups based on a filter attribute (e.g. memberURL).

Commit:
Update group_ldap.php

Commit:
Update configuration.php

PVince81 commented:
@blizzz can you have a look ?

blizzz commented:
Yes, will do

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -180,6 +184,34 @@ private function _groupMembers($dnGroup, &$seen = null) {
+													   $this->access->connection->ldapGroupFilter);
+			if ($memberURLs !== false) {
+				// this group has the 'memberURL' attribute so this is a dynamic group
+				$pos = strpos($memberURLs[0], "(");
```
please add an example  of the expected value in the comment, so a reviewer knows the format of the value.

alexweirig replied:
We've added examples, a simple and a more complex one, to show what the dynamic group attribute could look like (in this case it's the most common "memberURL" attribute).
More details can be found on this web page:
https://docs.oracle.com/cd/E19146-01/821-0793/geezc/index.html
We didn't want to put the URL in the code just to make sure we're not causing any issues because of the reference to an external web site.

blizzz replied:
Thanks!
URLs in comments are not a problem.

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -180,6 +184,34 @@ private function _groupMembers($dnGroup, &$seen = null) {
+
+		$dmEnabled = $this->access->connection->ldapUserFilterDMEnable;
+		$dmMemberURL = strtolower($this->access->connection->ldapUserFilterDMMemberUrl);
+		if ($dmMemberURL === '') {
```
should work, to be on the safe side please use ``empty($dmMemberURL)``.

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -180,6 +184,34 @@ private function _groupMembers($dnGroup, &$seen = null) {
+			$dmEnabled = false;
+		}
+
+		if ($dmEnabled) {
```
the method _groupMembers was already not short, so we should not make it longer. Please "outsource" this part to public getDynamicGroupMembers() which would also allow to more easily test this.

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -405,6 +440,12 @@ public function getUserGroups($uid) {
 
+		$dmEnabled = $this->access->connection->ldapUserFilterDMEnable;
+		$dmMemberURL = strtolower($this->access->connection->ldapUserFilterDMMemberUrl);
+		if ($dmMemberURL === '') {
```
empty($dmMemberURL)

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -422,11 +463,15 @@ public function getUserGroups($uid) {
 			
-			if($primaryGroup !== false) {
-				$groups[] = $primaryGroup;
+			if (!$dmEnabled) {
```
this and the following if clause can be combined. However, can you elaborate why using dynamic groups would skip fetching the primary group?

alexweirig replied:
I was actually maintaining the logic from the original code.  In the original code if a primary group was found then the function returned straight away.  Now in our case if dynamic membership is enabled we still do include the primary group and now the dynamic groups too.

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -458,6 +503,35 @@ public function getUserGroups($uid) {
 
+		if ($dmEnabled) {
+			// look through dynamic groups to add them to the result array if needed
+			$groupsToMatch = $this->access->fetchListOfGroups(
```
I wonder if its possible to extent this filter and take the memberUrl into account, in order to avoid the subsequent read operations in line 520?

alexweirig replied:
I have tried to see if this is possible but I do not think so.  As far as I can see there need to be two searches - one to get back the dynamic groups (i.e we are searching over group objects that have a 'memberURL' attribute) and then the second search to see if a given user belongs to that dynamic group.  This second search is searching over user objects so I do not see how the two searches can be combined into one search on the LDAP server.

blizzz replied:
Ok, I see. A pity, nevertheless thanks for looking into it.

Xenopathic commented on apps/user_ldap/group_ldap.php:
```diff
@@ -180,6 +184,34 @@ private function _groupMembers($dnGroup, &$seen = null) {
+
+		if ($dmEnabled) {
+			$memberURLs = $this->access->readAttribute($dnGroup, $dmMemberURL,
+													   $this->access->connection->ldapGroupFilter);
```
Multiple whitespace on one line for indentation. Please use only tabs. My recommendation is the following style:

```
$memberURLs = $this->access->readAttribute(
    $dnGroup,
    $dmMemberURL,
    $this->access->connection->ldapGroupFilter
);
```

This also has the advantage of keeping each parameter to an individual line, and keeps everything closer to the left margin (so smaller width editors don't have to scroll as far)

blizzz commented:
* [ ] In the configuration assistant, the setting does not fit into the users tab. The first four tabs are considered to be as sparse and necessary as possible to allow an easy configuration. This feature is a more advanced setup and should go to the Advanced Tab, Section "Directory Settings"
* [ ] Also in the configuration, combine the enabled checkbox and memberUrl setting into one text field. If the memberUrl-setting is entered by the admin user,  treat it as enabled, otherwise not. This will make it consistent with the other settings we have.

* In future, please use proper commit message that describe/summarize what the code changes are doing

Xenopathic commented on apps/user_ldap/js/wizard/wizardTabUserFilter.js:
```diff
@@ -55,7 +55,15 @@ OCA = OCA || {};
-				}
+				},
+				ldap_userfilter_dm_enable: {
+                                        $element: $('#ldap_userfilter_dm_enable'),
```
Whitespace issues

Xenopathic commented on apps/user_ldap/templates/part.wizard-userfilter.php:
```diff
@@ -55,6 +55,30 @@ class="ldapFilterInputElement"
+		</p>
+		<p>
+			<label>
+                        	Enable: 
```
Whitespace issues

alexweirig commented:
@Xenopathic @blizzz sorry for all the problems.
We'll review the code based on your comments

blizzz commented:
> @Xenopathic @blizzz sorry for all the problems.
We'll review the code based on your comments

Nothing to be sorry for. Reviewing is a standard process to keep up code quality and avoid issues. It's totally normal also amongst regular contributors to have some things that are requested to be changed.

Commit:
Revert to original version

This file should no longer be changed since we're putting the option to enable dynamic user groups onto a different page in wizard.

We simply don't know how to remove the file from the changes for this branch.

Commit:
Revert to original version

This file can be reverted to it's original content since we are putting the option to enable dynamic user groups into a different wizard page.

Commit:
Added support for dynamic user groups

Added attribute "ldapDynamicGroupMemberURL" to the $config array and added attribute "ldap_dynamic_group_member_url" to the array  returned by getDefaults and the mapping between the 2 fields in the array returned by getConfigTranslationArray.

Commit:
Added support for dynamic user groups

Added input field for ldap_dynamic_group_member_url in the Directory Settings section on the ldapAdvancedAccordion.

Commit:
Added support for dynamic user groups

Create the setDynamicGroupMemberURL method to support the field handling on the page

Commit:
Added support for dynamic user groups

The functions “getUserGroups” and “_groupMembers” have been adjusted and a new function “getDynamicGroupMembers” created to take into account dynamic group membership when this functionality has been enabled via the relevant settings on the LDAP→Advanced settings tab.

alexweirig commented:
@blizzz : We've updated the different code parts to meet your comments on the PR from yesterday.

We've also commented on some of the issues you raised.
We've taken care on the whitespace issues but we're not even sure why the appeared yesterday. Hopefully they're not related to the copy/paste of the code from the local editor to editor on the github website.

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -458,6 +513,35 @@ public function getUserGroups($uid) {
+					$memberUrlFilter = substr($value[$dynamicGroupMemberURL][0],$pos);
+					// apply filter via ldap search to see if this user is in this
+					// dynamic group
+					$userMatch = $this->access->readAttribute($uid, 'uid', $memberUrlFilter);
```
hardcoding the attribute is bad, because it might not be available in every setup. You can pass an empty string instead to check for the existence. Use the configured user display name attribute instead.

blizzz commented:
I believe  apps/user_ldap/js/wizard/wizardTabUserFilter.js was deleted by accident?

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -148,6 +148,46 @@ public function inGroup($uid, $gid) {
+			$pos = strpos($memberURLs[0], "(");
+			if ($pos !== false) {
+				$memberUrlFilter = substr($memberURLs[0], $pos);
+				$foundMembers = $this->access->searchUsers($memberUrlFilter,'uid');
```
also here, don't hardcode uid as attribute which might not be available, use the configured user display name attribute.

blizzz commented on apps/user_ldap/lib/configuration.php:
```diff
@@ -3,10 +3,10 @@
  * @author Alexander Bergolth <leo@strike.wu.ac.at>
  * @author Arthur Schiwon <blizzz@owncloud.com>
  * @author Jörn Friedrich Dreyer <jfd@butonic.de>
- * @author Lennart Rosam <hello@takuto.de>
```
why removed?

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -6,13 +6,13 @@
  * @author Bart Visscher <bartv@thisnet.nl>
  * @author Christopher Schäpers <kondou@ts.unde.re>
  * @author Frédéric Fortier <frederic.fortier@oronospolytechnique.com>
- * @author Lukas Reschke <lukas@owncloud.com>
```
why removed?

LukasReschke replied:
Nobody likes me :cry:

blizzz commented on apps/user_ldap/group_ldap.php:
```diff
@@ -148,6 +148,46 @@ public function inGroup($uid, $gid) {
+			// this group has the 'memberURL' attribute so this is a dynamic group
+			// example 1: ldap:///cn=users,cn=accounts,dc=dcsubbase,dc=dcbase??one?(o=HeadOffice)
+			// example 2: ldap:///cn=users,cn=accounts,dc=dcsubbase,dc=dcbase??one?(&(o=HeadOffice)(uidNumber>=500))
+			$pos = strpos($memberURLs[0], "(");
```
please use single quotes

alexweirig commented:
@LukasReschke sorry, we didn't remove those lines, it appears they were actually not in the code base we used when we applied the changes (8.2.1 release).
When we committed our code, we copy/pasted our code and thus must have by mistake removed those author lines.
Please accept my most sincerest apologies for having removed your name :-)

alexweirig commented:
we didn't know how to undo the changes we had initially made in /user_ldap/js/wizard/wizardTabUserFilter.js so we tried to "remove" the file from the pull request and thus must have removed it by mistake as you correctly noticed.

alexweirig commented:
Hello,

can we close or delete this pull request.

We've created a new one based on "latest" master.

Thanks 
Alex

alexweirig commented:
@blizzz here is a link to the new pull request https://github.com/alexweirig/core/pull/1

We created this pull request using the "github desktop" application.

Thanks for having a look at it and considering accepting our contribution.

Regards,

Alex

blizzz commented:
@alexweirig the pull request needs to be done against ownCloud/core, this one goes against your private repo. Could you reopen it once again? I would review it then this week.

alexweirig commented:
@blizzz I'm sorry I don't exactly understand what you mean with "reopen it once again"?

I've tried to work on owncloud/core and created a branch in the github desktop, then applied all the changes again in that branch and if I try to create a pull request, I'm getting a message that I'm not allowed to access owncloud/core????

This is getting somewhat frustrating, we've now tried to commit the code for the 3rd time ... and we're still not through with this github stuff.
Maybe we can create a teamviewer session and you can have a look at what's going on.
Things like this should be easy and straightforward in order to encourage people to participate, to me this is more of a pain ...

Commit:
adding support for dynamic ldap groups

The functions “getUserGroups” and “_groupMembers” have been adjusted and a new function “getDynamicGroupMembers” created to take into account dynamic group membership when this functionality has been enabled via the relevant settings on the LDAP→Advanced settings tab.

Commit:
adding support for dynamic ldap groups

Small updates to the existing configuration class to allow for the new setting defined above.

blizzz commented:
> I've tried to work on owncloud/core and created a branch in the github desktop, then applied all the changes again in that branch and if I try to create a pull request, I'm getting a message that I'm not allowed to access owncloud/core????

Yes – your dev branch is in your own repository. However, when you open the pull request, it needs to be done against the ownCloud repo, not yours. When you go to the webpage of your repo, and click on "New pull request", you should be able to pick "owncloud/core" as base fork repo (and keep "master" as it's branch). I have no idea about Github Desktop.

> This is getting somewhat frustrating, we've now tried to commit the code for the 3rd time ... and we're still not through with this github stuff.

Complex tools can be difficult at times, I am sorry about your frustration :( but we're almost there!

> Maybe we can create a teamviewer session and you can have a look at what's going on.
Things like this should be easy and straightforward in order to encourage people to participate, to me this is more of a pain ...

Once you know how this stuff works, it's pretty straightforward. We can do a brief session some time next week. I am pretty full Monday, Tue could work, depending on the time. Please email](https://github.com/owncloud/core/blob/master/apps/user_ldap/user_ldap.php#L3) me a suggestion and your time zone, and we make out something.

alexweirig commented:
@blizzz I see, now indeed I found it ... maybe this time it's OK

blizzz commented:
Closing in favor of https://github.com/owncloud/core/pull/21838
